### PR TITLE
Update signature for Expr.name so that schema is no longer required

### PR DESCRIPTION
**Pull request by andygrove**
# Which issue does this PR close?

Part of https://github.com/apache/arrow-datafusion/issues/3330

 # Rationale for this change

The schema argument to `Expr.name` was not used and not needed.

# What changes are included in this PR?

- Remove schema argument from `Expr.name`
- Remove some code duplication between `Display` and `Debug` traits

# Are there any user-facing changes?

Yes, API change.

### CODE REVIEW

File: datafusion/expr/src/expr.rs [hunk 1]
```diff
@@ -962,18 +942,31 @@ fn create_name(e: &Expr, input_schema: &DFSchema) -> Result<String> {
...
 
 #[cfg(test)]
 mod test {
     use crate::expr_fn::col;
-    use crate::lit;
+    use crate::{case, lit};
+    use datafusion_common::{Result, ScalarValue};
+
+    #[test]
+    fn format_case_when() -> Result<()> {
```
Comment (andygrove) [inline datafusion/expr/src/expr.rs:960]: I added this test just to demonstrate the difference between `Expr` `Display`, `Debug`, and `name()`. I plan on adding more tests in future PRs to cover other expressions.

File: datafusion/expr/src/expr.rs [hunk 2]
```diff
@@ -962,18 +907,31 @@ fn create_name(e: &Expr, input_schema: &DFSchema) -> Result<String> {
...
```
Comment (alamb) [inline datafusion/expr/src/expr.rs:925]: ❤️

File: datafusion/expr/src/utils.rs [hunk 3]
```diff
@@ -681,12 +681,7 @@ pub fn expr_as_column_expr(expr: &Expr, plan: &LogicalPlan) -> Result<Expr> {
             let field = plan.schema().field_from_column(col)?;
             Ok(Expr::Column(field.qualified_column()))
         }
-        _ => {
-            // we should not be trying to create a name for the expression
-            // based on the input schema but this is the current behavior
-            // see https://github.com/apache/arrow-datafusion/issues/2456
```
Comment (alamb) [inline datafusion/expr/src/utils.rs:687]: Looks like https://github.com/apache/arrow-datafusion/issues/2456  could be closed too, perhaps

**Review by alamb — APPROVED** [review]
Looks like a great improvement to me. Thanks @andygrove

**Pull request merged**